src/python/grpcio/grpc/_simple_stubs.py

```diff
@@ -0,0 +1,440 @@
+    channel_credentials = channel_credentials or grpc.local_channel_credentials(
+    )
+    if channel_credentials._credentials is grpc.experimental._insecure_channel_credentials:
+        _LOGGER.info(f"Creating insecure channel with options '{options}' " +
+                     f"and compression '{compression}'")
+        return grpc.insecure_channel(target,
+                                     options=options,
+                                     compression=compression)
+    else:
+        _LOGGER.info(
+            f"Creating secure channel with credentials '{channel_credentials}', "
+            + f"options '{options}' and compression '{compression}'")
+        return grpc.secure_channel(target,
+                                   credentials=channel_credentials,
+                                   options=options,
+                                   compression=compression)
+
+
+class ChannelCache:
+    _singleton = None
+    _lock = threading.RLock()
+    _condition = threading.Condition(lock=_lock)
+    _eviction_ready = threading.Event()
+
+    def __init__(self):
+        self._mapping = collections.OrderedDict()
+        self._eviction_thread = threading.Thread(
+            target=ChannelCache._perform_evictions, daemon=True)
+        self._eviction_thread.start()
+
+    @staticmethod
+    def get():
+        with ChannelCache._lock:
+            if ChannelCache._singleton is None:
+                ChannelCache._singleton = ChannelCache()
+        ChannelCache._eviction_ready.wait()
+        return ChannelCache._singleton
+
+    # TODO: Type annotate key.
+    def _evict_locked(self, key):
+        channel, _ = self._mapping.pop(key)
+        _LOGGER.info("Evicting channel %s with configuration %s.", channel, key)
+        channel.close()
+        del channel
+
+    # TODO: Refactor. Way too deeply nested.
+    @staticmethod
+    def _perform_evictions():
+        while True:
+            with ChannelCache._lock:
+                ChannelCache._eviction_ready.set()
+                if not ChannelCache._singleton._mapping:
+                    ChannelCache._condition.wait()
+                elif len(ChannelCache._singleton._mapping) > _MAXIMUM_CHANNELS:
+                    key = next(iter(ChannelCache._singleton._mapping.keys()))
+                    ChannelCache._singleton._evict_locked(key)
+                    # And immediately reevaluate.
+                else:
+                    key, (_, eviction_time) = next(
+                        iter(ChannelCache._singleton._mapping.items()))
+                    now = datetime.datetime.now()
+                    if eviction_time <= now:
+                        ChannelCache._singleton._evict_locked(key)
+                        continue
+                    else:
+                        time_to_eviction = (eviction_time - now).total_seconds()
+                        ChannelCache._condition.wait(timeout=time_to_eviction)
```

Comment:
optional: when the eviction thread is waiting in the eviction period, the `_mapping` could exceed `_MAXIMUM_CHANNELS`. One alternative is that each eviction cycle, it can remove any number of expired channels, so both the number of channels is controlled and eviction time is respected.

You could add a test case that the application creates many channels within one eviction period.

Reply:
Discussed offline.

This is true. However, the thread calling `get_channel` notifies when it creates a channel and we are at or above `_MAXIMUM_CHANNELS`. So the eviction thread will soon wake up and rectify the situation.

As for the test you suggest, I believe I already wrote one. It caught several race conditions.